[MRG+1] Add multiplicative-update solver in NMF, with all beta-divergence

TomDLT opened the pull request:
This PR is a second part of what have been discussed in #4811. (first part, #4852, merged)
It includes :
- a Multiplicative-Update solver in NMF. This solver is generally slower than the Coordinate-Descent solver, but it handles all beta-divergences (including Frobenius, generalized Kullback_Leibler and Itakura-Saito). 
- a plot to visualize the beta-divergence for several values of beta.
- benchmarks [below](https://github.com/scikit-learn/scikit-learn/pull/5295#issuecomment-174002806)

[Link to the rendered doc](https://5075-843222-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/decomposition.html#nmf-with-a-beta-divergence)

---

plot_beta_divergence.py
![beta_div](https://cloud.githubusercontent.com/assets/11065596/10021621/232917e2-6149-11e5-84f6-8e9b09578216.png)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/5295)

<!-- Reviewable:end -->

mblondel commented:
I would favor using CD for beta divergences as well. We decided to remove the slow pg solver. I would prefer not add a new slow solver if possible. The KDD paper shows that for generalized KL, we can just use the element-wise Newton update without line search. For other divergences, this will need some verification (unless there are more recent works that cover this?) but in worst case we can implement a simple line search. As a start, it would be nice to implement the CD update for generalized KL to compare.

tttthomasssss commented:
Could anybody comment on when this is planned to be merged into master? I'm quite keen on using NMF with generalized KL divergence.

agramfort commented:
@mblondel maybe we can do better than multiplicative updates but I think we
should be pragmatic here. We can deprecate after when we have something
better. wdyt?

TomDLT commented:
Sorry for the long silence, I have not as much time for this as before.
@mblondel I spent some time trying to implement the coordinate descent for generalized Kullback Leibler divergence, as described in [Coordinate descent with greedy selection](http://www.cs.utexas.edu/~cjhsieh/nmf_kdd11.pdf), at least in the dense case.
However, I did not manage to be faster than multiplicative update, which is not what is indicated in the paper, and probably underlines my own limits (or proves that my implementation of multiplicative update is very good :)). More seriously, even if we manage to be faster, the extension to all beta-divergence is not trivial and would recquire much more work.

I understand that multiplicative update is an old method and probably not the fastest, yet it is still a good way to improve the NMF in scikit-learn, extending to a lot of different loss functions.

TomDLT commented:
I compared my implementation of multiplicative update with #2540 (Frobenius norm) and #1348 (KL divergence).
___
Comparing to #2540, the results are identical, provided some slight modifications:
To avoid division by zero, what do we prefer:
1. `A / (B + eps)` (like in #2540)
2. `(A + eps) / (B + eps)` (like in #1348)
3. `B[B == 0] = eps` then `A / B`
I have a preference for the last one, for it does not affect update with nonzero elements. WDYT ?
___
Comparing to #1348, the results are different since it forgets a division in the update
(see Eq 5 [here](http://papers.nips.cc/paper/1861-algorithms-for-non-negative-matrix-factorization.pdf)), and adds a normalization of H.

TomDLT commented:
**Some benchmarking:**
20_news - sparse (11314, 39116)
![20news](https://cloud.githubusercontent.com/assets/11065596/12518151/0693278c-c137-11e5-8c25-fdf86f0fe7e4.png)
Faces - dense (400, 4096)
![faces](https://cloud.githubusercontent.com/assets/11065596/12518150/06927e5e-c137-11e5-875e-95c29ed558a9.png)
MNIST - dense (70000, 784)
![mnist](https://cloud.githubusercontent.com/assets/11065596/12518153/06a62f76-c137-11e5-94be-23a14e80fe29.png)
RCV1 - sparse (804414, 47236)
![rcv1](https://cloud.githubusercontent.com/assets/11065596/12518152/06a4c76c-c137-11e5-9e8b-4ea75f651130.png)

The MU solver is not very fast, but it was implemented not for its speed, but for it handles all beta-divergences. Note that the poor performances with 'nndsvd' initialization (top-left) are expected, since this initialization have a lot of zeros that cannot be modified by a Multiplicative Update.

TomDLT committed:
ENH add multiplicative-update solver in NMF, with all beta-divergence

tguillemot commented:
@TomDLT @mblondel @agramfort This PR is mergeable ? 
I will do a first round of review but I don't know those methods. Someone can have a look with me ?

tguillemot commented on benchmarks/bench_plot_nmf.py:
```diff
@@ -1,166 +1,162 @@
-
-    return timeset, err
+import matplotlib.pyplot as plt
+import pandas
```
We can use pandas for benchs ?

TomDLT replied:
Using it in a benchmark does not make it a dependency of the package, as for matplotlib.

tguillemot commented on sklearn/decomposition/nmf.py:
```diff
@@ -78,19 +81,156 @@ def _safe_compute_error(X, W, H):
+    If beta == 1, this is the generalized Kullback-Leibler divergence
+    If beta == 0, this is the Itakura-Saito divergence
+    Else, this is the general beta-divergence.
+    """
```
Can you specify what are X, W, H, and beta with a classical docstring

tguillemot commented on sklearn/decomposition/nmf.py:
```diff
@@ -78,19 +81,156 @@ def _safe_compute_error(X, W, H):
+            # np.sum(np.dot(W, H) ** beta)
+            sum_WH_beta = 0
+            for i in range(X.shape[1]):
+                sum_WH_beta += np.sum(fast_dot(W, H[:, i]) ** beta)
```
Is there a better option ?

TomDLT replied:
This is a temporary fix so as to be memory efficient in the sparse case.
(We cannot compute directly `np.dot(W, H)`).

We could have a cython code for it, yet the use case is rather limited (sparse X, and beta not in [0, 1, 2] which are the most useful cases), so I propose to merge it as it is, and improve it in a separate PR.

tguillemot replied:
Good for me :)

agramfort commented:
@mblondel what's your take on this PR?

mblondel commented:
> I spent some time trying to implement the coordinate descent for generalized Kullback Leibler divergence, as described in Coordinate descent with greedy selection, at least in the dense case.

Thanks for the investigation.

> More seriously, even if we manage to be faster, the extension to all beta-divergence is not trivial and would require much more work.

Agreed.

> @mblondel what's your take on this PR?

+1 on my side and sorry for the late reply. As a big fan of CD, I would be potentially interested in the CD code for GKL divergence as it could be faster in the sparse case.

agramfort commented:
cool. Looks like we can proceed then :)

let's make it a WIP->MRG then

TomDLT commented:
> let's make it a WIP->MRG then

it is :wink:, reviews will be greatly appreciated.

TomDLT committed:
improve docstring

tguillemot commented on doc/modules/decomposition.rst:
```diff
@@ -724,10 +723,57 @@ and the regularized objective function is:
+However, NMF can also be used with a different function to measure the
+distance between X and the matrix product WH. Another typical distance
+function used in NMF is the (generalized) Kullback-Leibler (KL) divergence,
+also referred as I-divergence:
```
I would change a bit the formulation :
"Other distance functions can be used in NMF as, for example, the (generalized) Kullback-Leibler (KL) divergence, also referred as I-divergence:

..math::
     d_{KL}(X, Y) = \sum_{i,j} (X_{ij} * log(\frac{X_{ij}}{Y_{ij}}) - X_{ij} + Y_{ij})

Or, the the Itakura-Saito (IS) divergence:"
See if you prefer.

tguillemot commented on doc/modules/decomposition.rst:
```diff
@@ -724,10 +723,57 @@ and the regularized objective function is:
+    d_{IS}(X, Y) = \sum_{i,j} (\frac{X_{ij}}{Y_{ij}} - log(\frac{X_{ij}}{Y_{ij}}) - 1)
+
+These three distances are special cases of the beta-divergence family, with
+:math:`\beta = 2, 1, 0` respectively [Fevotte, 2011]. The beta-divergence are
```
Maybe you can add a link to [Fevotte, 2011] (even if it's on the references section)

tguillemot commented on doc/modules/decomposition.rst:
```diff
@@ -724,10 +723,57 @@ and the regularized objective function is:
+also referred as I-divergence:
+
+.. math::
+    d_{KL}(X, Y) = \sum_{i,j} (X_{ij} * log(\frac{X_{ij}}{Y_{ij}}) - X_{ij} + Y_{ij})
```
Remove the multiplication symbol `*`

TomDLT committed:
DOC Add links to references

TomDLT committed:
FIX link to reference

TomDLT commented:
Thanks for the comments @tguillemot, I addressed them.

[Link to the rendered doc](https://5071-843222-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/decomposition.html#nmf-with-a-beta-divergence)

ogrisel commented on doc/modules/decomposition.rst:
```diff
@@ -686,7 +685,7 @@ faces dataset, in comparison with the PCA eigenfaces.
 The :attr:`init` attribute determines the initialization method applied, which
 has a great impact on the performance of the method. :class:`NMF` implements
-the method Nonnegative Double Singular Value Decomposition. NNDSVD is based on
+the method Nonnegative Double Singular Value Decomposition. NNDSVD [4]_ is based on
```
Maybe explain why the NNDSVD init should not be used with the multiplicative updates solver at this end of this paragraph.

TomDLT replied:
Good point

ogrisel commented on sklearn/decomposition/nmf.py:
```diff
@@ -80,19 +83,166 @@ def _safe_compute_error(X, W, H):
+        raise ValueError(
+            'Invalid beta_loss parameter: solver %r does not handle beta_loss'
+            ' = %r' % (solver, beta_loss))
+
```
Maybe issue a user warning when the mu solver is used in conjunction with the nndsvd init.

TomDLT replied:
Done

ogrisel reviewed (commented):
How do you explain that MU behaves so badly with NNDSVDAR init on 20newsgroups? Does it changes significantly if you change the random seed?

TomDLT committed:
add warning when solver='mu' and init='nndsvd'

TomDLT committed:
revert all change in benchmark (separate PR)

TomDLT committed:
add example in the doc and docstring

TomDLT committed:
change versionadded to 0.19

TomDLT committed:
fix doctest

TomDLT commented:
Interestingly, I relaunched my benchmark without changes, and MU is not stuck anymore in a local minimum as previously with NDDSVDAR init. Or rather all three solvers are stuck in this local minimum at 103.205 instead of 103.200.

Also the random inits now give poorer convergence speed for all solvers. I don't know what changed.

Before
![20news](https://cloud.githubusercontent.com/assets/11065596/12518151/0693278c-c137-11e5-8c25-fdf86f0fe7e4.png)
After
![20_news__](https://cloud.githubusercontent.com/assets/11065596/18706993/66ea7c1c-7ff4-11e6-8dd8-d3dbc4c721fd.png)

ogrisel commented:
Good for 20 newsgroups.

With respect to https://github.com/scikit-learn/scikit-learn/pull/5295#issuecomment-158010088 I also prefer #3.

ogrisel reviewed (changes requested):
I have not yet finished my review but are some more comments. Also I found the following case where mu seems to underperform:

``` python
import numpy as np
from sklearn.decomposition import NMF

rng = np.random.RandomState(42)
n_samples, n_features = 500, 50
k = 10
n_components = min(2 * k, n_features)

W_true = np.clip(rng.randn(n_samples, k), 0, np.inf)
H_true = np.clip(rng.randn(k, n_features), 0, np.inf)
X = np.dot(W_true, H_true)


def nmf_error(X, **model_params):
    nmf = NMF(**model_params)
    W = nmf.fit_transform(X)
    H = nmf.components_
    return np.linalg.norm(X - np.dot(W, H))


for solver in ['cd', 'mu', 'pg']:
    error = nmf_error(X, n_components=n_components, solver=solver,
                      random_state=0, init='nndsvdar', max_iter=500)
    print('solver=%s, error=%f' % (solver, error))
```

with output:

```
solver=cd, error=0.114302
solver=mu, error=3.146888
/volatile/ogrisel/code/scikit-learn/sklearn/decomposition/nmf.py:1217: DeprecationWarning: 'pg' solver will be removed in release 0.19. Use 'cd' solver instead.
  nls_max_iter,
/volatile/ogrisel/code/scikit-learn/sklearn/decomposition/nmf.py:496: UserWarning: Iteration limit reached in nls subproblem.
  warnings.warn("Iteration limit reached in nls subproblem.")
solver=pg, error=0.506917
```

Any idea why? I need to play more with the rng seed an problem size but it might be a bug no?

TomDLT commented on sklearn/decomposition/nmf.py:
```diff
@@ -80,19 +83,173 @@ def _safe_compute_error(X, W, H):
+
+
+def _compute_regularization(alpha, l1_ratio, regularization, n_samples,
+                            n_features):
```
This is related to #5296, yet I will remove them in this PR.

TomDLT commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,9 +741,264 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+        else:
+            WH_safe_X_data **= beta_loss - 2
+            # element-wise multiplication
+            WH_safe_X_data *= X_data
```
I already separated the case `beta_loss == 2` just above.

ogrisel replied:
Indeed, sorry for the noise.

TomDLT commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,9 +741,264 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+            H *= delta_H
+
+            # These values will be recomputed since H changed
+            H_sum, HHt, XHt = None, None, None
```
In #1348, the convergence check is `prev_error - error < tol`.
In #2540, the convergence check is `abs(1. - max_update) < tol`.

About computing the error, the cost is a bit high, yet I need to test it again.

About checking the maximum update, the update indeed converges to 1, but the criterion is rather poor if I recall correctly. After discussing with audio researchers in my group that use NMF a lot, a fixed number of iteration seems pretty classic.
Yet I can easily add this criterion if we consider it a better solution.

TomDLT replied:
Computing the error at each iteration increases the computational time by +50%.
(Tested on 20news dataset, with several beta losses)

ogrisel replied:
Let's do it every 10 iterations then.

TomDLT commented on sklearn/decomposition/nmf.py:
```diff
@@ -28,6 +29,8 @@
 from ..exceptions import ConvergenceWarning
 from .cdnmf_fast import _update_cdnmf_fast
 
+EPSILSON = 1e-9
```
Done

TomDLT commented on doc/modules/decomposition.rst:
```diff
@@ -724,35 +728,98 @@ and the regularized objective function is:
+defined by :
+
+.. math::
+    d_{\beta}(X, Y) = \sum_{i,j} \frac{1}{\beta(\beta - 1)}(X_{ij}^\beta + (\beta-1)Y_{ij}^\beta - \beta X_{ij} Y_{ij}^{\beta - 1})
```
Done

TomDLT commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,9 +741,264 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+            # copy used in the Denominator
+            WH = WH_safe_X.copy()
+
+        # to avoid division by zero
```
Done

TomDLT committed:
address review's comments

TomDLT commented:
About your toy example, increasing the number of iteration solves the problem, yet I don't know why it is that slow.

TomDLT committed:
Temporary: test a stopping criterion in nmf-MU

ogrisel commented:
> About your toy example, increasing the number of iteration solves the problem

Indeed, I thought I had tried that but maybe I was in slowly decreasing / plateau like area of the convergence curve and I had not tried to increase max_iter enough at once.

> yet I don't know why it is that slow.

That's weird but apparently it does not seem to be so bad on more realistic data as your benchmarks show.

ogrisel commented:
> About computing the error, the cost is a bit high, yet I need to test it again.

Then we could compute the cost every 10 or 100 iter then.

ogrisel commented:
> About checking the maximum update, the update indeed converges to 1, but the criterion is rather poor if I recall correctly.

I tried to add print statements to monitor the maximum absolute values in the delta multipliers and it seems that at least for some components we can get into an oscillatory regime.

ogrisel commented:
Also ideally we would like that the `tol` based criterion to be invariant to scaling of X.

TomDLT committed:
update convergence criterion and tests to avoid warnings

TomDLT commented:
The convergence test is now done every 100 iterations, and is normalized:
```python
if (previous_error - error) / error < tol:
    break
```

ogrisel commented:
```python
if (previous_error - error) / error < tol:
    break
```

I am afraid this can diver on problems where error can get very close to zero very suddenly. Maybe the following is safer:

```python

# difference of errors normalized by mean error across the last two
# iterations.
if  2 * (previous_error - error) / (previous_error + error) < tol:
    break
```

or even:


```python
if  previous_error - error / previous_error < tol:
    break
```

if we have the guarantee that the error is monotonic.

ogrisel commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,9 +740,287 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+
+    if verbose:
+        end_time = time.time()
+        print("Epoch %02d reached after %.3f seconds; convergence reached." %
```
Please remove "; convergence reached" from this message.

ogrisel commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,9 +740,287 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+            if verbose:
+                epoch_time = time.time()
+                print("Epoch %02d reached after %.3f seconds." %
+                      (epoch + 1, epoch_time - start_time))
```
Please include the value of the objective function in this message.

ogrisel commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,9 +740,287 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+        if tol > 0 and epoch % 100 == 0:
+            error = beta_divergence(X, W, H, beta_loss)
+            if (previous_error - error) / error < tol:
+                break
```
Actually this is too strict and never works on some toy problem I am testing on. I thought about storing the initial value of the error in a variable called `error_at_init` and then checking for:

```python
if (previous_error - error) / error_at_init < tol:
    break
```

but then it's hard to define a good value for tol.

TomDLT commented:
Good point. I updated with your last suggestion, as the error is indeed monotonic.

```python
if (previous_error - error) / previous_error < tol:
    break
```
I also updated the tests to avoid or ignore `ConvergenceWarning`.

ogrisel commented:
Here is what I tried, not sure this is the best stopping criterion normalization though:

```
diff --git a/sklearn/decomposition/nmf.py b/sklearn/decomposition/nmf.py
index 47b934d..8e1d1d3 100644
--- a/sklearn/decomposition/nmf.py
+++ b/sklearn/decomposition/nmf.py
@@ -974,10 +974,10 @@ def _fit_multiplicative_update(X, W, H, beta_loss='frobenius',
     else:
         gamma = 1.
 
+    start_time = time.time()
     # used for the convergence criterion
-    previous_error = np.inf
+    error_at_init = previous_error = beta_divergence(X, W, H, beta_loss)
 
-    start_time = time.time()
     H_sum, HHt, XHt = None, None, None
     for epoch in range(max_iter):
         # update W
@@ -999,14 +999,14 @@ def _fit_multiplicative_update(X, W, H, beta_loss='frobenius',
         # test convergence criterion every 100 iterations
         if tol > 0 and epoch % 100 == 0:
             error = beta_divergence(X, W, H, beta_loss)
-            if (previous_error - error) / error < tol:
+            if (previous_error - error) / error_at_init < tol:
                 break
             previous_error = error
 
             if verbose:
                 epoch_time = time.time()
-                print("Epoch %02d reached after %.3f seconds." %
-                      (epoch + 1, epoch_time - start_time))
+                print("Epoch %02d reached after %.3f seconds, error: %f" %
+                      (epoch + 1, epoch_time - start_time, error))
 
     if verbose:
         end_time = time.time()
```

ogrisel commented:
In my experiments `(previous_error - error) / previous_error < tol` can also be too strict.

ogrisel commented:
> I also updated the tests to avoid or ignore ConvergenceWarning.

We should also make sure that the tol / criterion we use by default make all example run without issuing a `ConvergenceWarning`.

TomDLT committed:
normalize convergence criterion with error_at_init

TomDLT commented:
I updated the normalization with `error_at_init` which seems indeed a good idea.

I did not update examples, as they all use the default solver (`solver='cd'`).

ogrisel commented:
> I did not update examples, as they all use the default solver (solver='cd').

That means that we need an example to highlight the use of the new beta divergences. Any suggestion for a dataset (preferably already in scikit-learn) where they make a qualitative difference?

ogrisel commented:
I also think we should increase the default value of `max_iter` to 1000.

ogrisel commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -258,3 +289,225 @@ def test_safe_compute_error():
+    rng = np.random.mtrand.RandomState(42)
+    X = rng.randn(n_samples, n_features)
+    np.abs(X, X)
+    W, H = nmf._initialize_nmf(X, n_components, init='random', random_state=42)
```
This should be called within the inner for loop no?

ogrisel replied:
I mean just before `previous_loss =None`.

TomDLT replied:
There was indeed a bug.
I added a copy of `W0` and `H0`.

ogrisel commented:
Could you also please re-run the benchmarks with the new stopping criterion to check that it's behaving correctly on those datasets and that the overhead is not too costly?

TomDLT committed:
Fix test adding a copy of shared inititalization

TomDLT commented:
> That means that we need an example to highlight the use of the new beta divergence

It can be shown that the NMF with Kullback-Leibler generalized divergence is equivalent [1] to Probabilistic Latent Semantic Indexing, a popular method for document clustering, also related to the Latent Dirichlet Allocation.
We could extend `topics_extraction_with_nmf_lda.py` with the NMF using `beta_loss='kullback-leibler'`, WDYT?

[1] [Ding, C., Li, T., & Peng, W. (2008). On the equivalence between non-negative matrix factorization and probabilistic latent semantic indexing. Computational Statistics & Data Analysis, 52(8), 3913-3927.](http://users.cis.fiu.edu/~taoli/pub/NMFpLSIequiv.pdf)

> I also think we should increase the default value of max_iter to 1000.

For all solvers? Do we need to deprecate the old default value?

TomDLT committed:
add NMF with KL divergence in topic extraction example

ogrisel commented:
> We could extend topics_extraction_with_nmf_lda.py with the NMF using beta_loss='kullback-leibler', WDYT?

+1.

ogrisel commented:
>> I also think we should increase the default value of max_iter to 1000.
> For all solvers? Do we need to deprecate the old default value?

It depends. If mu converges with the default value of tol and max_iter on the 20 newsgroups topic example then let's keep `max_iter=200`.

If `max_iter` proves too small the typical use case then I think that's a usability problem. Having a default value for `max_iter` that is different for different solvers is counter intuitive. If we decide to change the value of `max_iter` for all solvers then introducing a `ChangedBehaviorWarning` is a possibility although I am not sure it's necessary.

ogrisel commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -502,10 +503,10 @@ def test_nmf_decreasing():
+                    X, W0.copy(), H0.copy(), beta_loss=beta_loss,
+                    n_components=n_components, max_iter=n_iter, alpha=alpha,
+                    solver=solver, tol=tol, l1_ratio=l1_ratio, verbose=0,
+                    regularization='both', random_state=0, update_H=True)
```
I would instead insert:

```python
W, H = W0.copy(), H0.copy()
previous_loss = None
for n_iter in range(1, 5):
    W, H, n_iter = non_negative_factorization(X, W, H, ...)
```

so as to check the monotonicty insude the `n_iter` loop.

TomDLT replied:
I changed it according to your suggestion, changing to `n_iter=1` since we want only one more iteration starting from the previous results.

It made me realize I forgot `init='custom'` :worried:...

TomDLT committed:
Fix add init parameter for custom init

TomDLT committed:
decrease to 10 iteration between convergence test

TomDLT commented:
I just realized that by testing only every 100 iterations, we aggregate the changes across 100 iterations, which makes us wait the convergence much longer than needed.

I changed it to check only every 10 iterations.
It converges at iteration 30 to 60 on 20news dataset, depending on the initialization.
It converges at iteration 220 to 420 on faces dataset. depending on the initialization.

TomDLT committed:
Fix the reconstruction error from x**2 / 2 to x

TomDLT committed:
fix init docstring

TomDLT commented:
Any idea why Travis is raising a PEP8 error in the `setup.py` file ?

lesteve commented:
@TomDLT I merged #7575. I'll trigger the rebuild of the flake8 build to see whether that fixes it.

lesteve commented:
> @TomDLT I merged #7575. I'll trigger the rebuild of the flake8 build to see whether that fixes it.

Yep that fixed it, great way to sanity check the fix!

TomDLT commented:
Good job @lesteve , thanks !

ogrisel commented on sklearn/decomposition/tests/test_nmf.py:
```diff
@@ -258,3 +287,228 @@ def test_safe_compute_error():
+                continue
+            W, H = W0.copy(), H0.copy()
+            previous_loss = None
+            for _ in range(1, 5):
```
I would rather make the test stricter with:

```python
for _ in range(10):
    ...
```

ogrisel replied:
It still runs in 800ms.

TomDLT replied:
I added `nls_max_iter=20` since the bottleneck is the 'pg' solver.
With this option, we can go up to 30 iterations in 300ms.

ogrisel reviewed (approved):
Besides the following small comments, +1 for merge.

TomDLT committed:
typo and improve test decreasing

TomDLT commented:
Final comments addressed. Thanks a lot for the review @ogrisel !

TomDLT committed:
remove unused private function _safe_compute_error

ogrisel commented:
Once this is merged, #5779 should be refactored to benchmark against this implementation before merging it.

amueller reviewed (commented):
If the pg solver will be removed in 0.19, we can just drop it now, right?

TomDLT committed:
make beta_divergence function private

TomDLT committed:
Remove deprecated ProjectedGradientNMF

TomDLT committed:
remove warning in test

TomDLT committed:
update doc

ogrisel commented:
> If the pg solver will be removed in 0.19, we can just drop it now, right?

+1

TomDLT committed:
FIX raise an error when beta_loss <= 0 and X contains zeros

TomDLT committed:
TYPO epsilson -> epsilon

TomDLT commented:
@amueller I addressed your comments, and **removed completely the projected gradient solver**.
___
I also realized that for a negative (or zero) `beta_loss`, the beta divergence of `(x, y)` is negative when `x = 0`, and the beta divergence of `(0, y)` goes to `-inf` as `y` goes to zero, which breaks the optimization. 

So I added an error when `beta_loss <= 0` and `X` contains zeros, I added a test and updated the doc and docstrings.

TomDLT committed:
remove other occurences of ProjectedGradientNMF

TomDLT committed:
add whats_new.rst entry

TomDLT commented:
I rebased due to conflicts while removing last `ProjectedGradientNMF` occurences.
I added a whats_new.rst entry

TomDLT committed:
minor leftovers

ogrisel commented on sklearn/decomposition/nmf.py:
```diff
@@ -600,12 +500,309 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+
+        # necessary for stability with beta_loss < 1
+        if beta_loss < 1:
+            W[W < 1e-15] = 0.
```
better use `np.ffinfo(np.float64).eps` no?

ogrisel reviewed (commented):
I cannot find the non-ascii char reported by the test failure but I found I found the following nitpick instead:

TomDLT committed:
non-ascii and nitpick

TomDLT committed:
safe_min instead of min

TomDLT commented:
Done

amueller commented:
It looks good but I didn't (and wont) have time to review the math.

amueller commented:
needs a rebase :-/

TomDLT commented:
Done

amueller commented:
Thanks. Hm so @mblondel I saw a +1 from you up there, but that wasn't a full review, was it? @agramfort did you review? I see +1 from @ogrisel

agramfort commented:
no I did not. No time :(

VictorBst commented:
Hi,  @TomDLT  and @agramfort invited to have a look at the code.

I mostly looked at the math aspect of the code and it seems to be correct. After quick experiments on my data, everything seems to work as intended and gives similar results to other simple NMF python codes I tested. As a side note, this one was always either equal or the fastest in my tests, especially for higher dimensions.

TomDLT committed:
solve conflict with master

amueller commented:
hm I'm not sure if this should be merged before #7927, there are some conflicts, I think.... opinions @TomDLT ?

TomDLT commented:
ok let's wait for #7927

ogrisel commented:
@TomDLT now that #7927 has been merged, this needs a rebase.

TomDLT committed:
Merge branch 'master' into nmf_mu

TomDLT committed:
minor doc update

TomDLT commented:
Done

ogrisel commented:
Thanks @TomDLT (and @VictorBst for the review). I squash-merged. :beers:

TomDLT commented:
:champagne: Cool ! Thanks a lot for the reviews ! :champagne:

tguillemot commented:
Yeah !!!

amueller commented:
Awesome! Congrats!
